sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubMessageToRow.java

```diff
@@ -90,44 +102,94 @@ public static Builder builder() {
     return new AutoValue_PubsubMessageToRow.Builder();
   }
 
-  @DoFn.ProcessElement
-  public void processElement(ProcessContext context) {
-    try {
-      List<Object> values = getFieldValues(context);
-      context.output(Row.withSchema(messageSchema()).addValues(values).build());
-    } catch (UnsupportedRowJsonException jsonException) {
-      if (useDlq()) {
-        context.output(DLQ_TAG, context.element());
-      } else {
-        throw new RuntimeException("Error parsing message", jsonException);
-      }
+  private PubsubIO.Write<PubsubMessage> writeMessagesToDlq() {
+    PubsubIO.Write<PubsubMessage> write = PubsubIO.writeMessages().to(deadLetterQueue());
+
+    return timestampAttribute() != null
+        ? write.withTimestampAttribute(timestampAttribute())
+        : write;
+  }
+
+  @Override
+  public PCollection<Row> expand(PCollection<PubsubMessage> input) {
+    Schema schema = payloadSchema();
+    PCollectionTuple rows =
+        input.apply(
+            ParDo.of(
+                    useFlatSchema()
+                        ? new FlatSchemaPubsubMessageToRoW(schema)
+                        : new NestedSchemaPubsubMessageToRow(schema))
```

Comment:
It would be nice if this were the only place that we branch based on `useFlatSchema` (currently it also happens inside `payloadSchema()`). 

What do you think about passing the message schema in the constructor for these DoFn classes instead? Then each implementation can be responsible for extracting the payload schema they need, we can get rid of `payloadSchema()`, and the `DoFn`s can be static (I think).

Reply:
Done.